### ISSUE TITLE
Send language server stderr to a file

### ISSUE THREAD

**Issue by sam-mccall** [neovim/neovim#16807]
### Feature description

TL;DR: as a language server maintainer, it would be useful if the language server stderr stream was written to a temporary file. It's okay if this is optional.

Without this:
 - nvim users can't usefully report bugs against our server
 - it's painful to understand/debug server interactions with nvim

---

Background: clangd and many other language servers produce detailed logs on stderr by default.
These logs have all critical information for investigating server behavior, e.g. protocol messages interleaved with internal state, crash dumps etc.

We require logs with bug reports, and provide [per-editor instructions](https://clangd.llvm.org/installation.html) for how to obtain them.
Here's [an example bug](https://github.com/clangd/clangd/issues/859) investigated with the help of logs. 
Viewing logs is also a critical part of development, long live printf debugging!

---

Today nvim sends the error stream to the lsp.log file in the following format:

```
[ERROR][2021-12-25 23:48:54] .../vim/lsp/rpc.lua:417	"rpc"	"clangd"	"stderr"	"I[23:48:54.673] clangd version 14.0.0 (48d868493fa74025e7768afacdbbbd3ea9c82468)\nI[23:48:54.673] Features: linux+grpc\nI[23:48:54.673] PID: 27944\nI[23:48:54.673] Working directory: /home/sammccall\nI[23:48:54.673] argv[0]: clangd\nI[23:48:54.673] argv[1]: --background-index\n"
```

This isn't useful because:
 - it's hard to read due to the long prefixes and escaping
 - structure isn't visible because line boundaries aren't respected (lines are both combined and split)
 - it combines logs from several instances of the server, so:
   - it's too hard to trace state if the instances are concurrent
   - users don't know where to cut logs (it's *so* hard to get them to provide full logs aready)
   - it's easy for users to leak details of private code into a bug report for an isolated example

---

What I'd like is for the stderr stream to be sent to a path created by tempname(), and for this path to be logged to lsp.log and/or available programmatically.

This means the file would be scoped to the lifetime of the editor - it would remain if clangd exits/crashes but be removed once the editor is closed.

I've implemented this solution in my own vim LS-client from years ago and it works really well.

(Passing a --log-file flag to the server isn't a great solution: between ensuring unique file paths, discoverability of the file paths, and giving the files the right lifetime, it's hard to get this right without editor cooperation)

**Comment by mjlbach** [comment]
It is sent to a file, lsp.log, prefixed with [error]

**Comment by mjlbach** [comment]
> it's hard to read due to the long prefixes and escaping

require('vim.lsp.log').set_format_func(vim.inspect) helps quite a bit, I've considered making it the default

> structure isn't visible because line boundaries aren't respected (lines are both combined and split)
    it combines logs from several instances of the server, so:
        it's too hard to trace state if the instances are concurrent
        users don't know where to cut logs (it's so hard to get them to provide full logs aready)
        it's easy for users to leak details of private code into a bug report for an isolated example

I'm a little wary of splitting the logs (I'm not saying no). It's pretty trivial to write a short lua script that will pull all [error] prefixed errors out of `lsp.log`, and you can ask your users to clear lsp.log before recording the issue.

**Comment by sam-mccall** [comment]
> How does vscode handle this? You work on coc-clangd and vscode-clangd, so I assume it is the temp file solution you suggested?

VSCode lets you connect server output to a named "output channel", which is either an in-memory or on-disk buffer (I'm not sure), and it's nice and responsive. It's selectable in the output panel like this: ![](https://i.stack.imgur.com/Cpmqi.png)

coc.nvim buffers the output somewhere (again I'm not sure where) and you can show it with `:CocCommand workspace.showOutput`. (If I've had a long session with verbose, this tends to spin my CPU forever. I think they're sending the whole log content from node->vim over the IPC...)

(coc-clangd and vscode-clangd are thin wrappers analogous to nvim-lspconfig, so they just follow the convention of the main editor/plugin. The client I mentioned working on was a separate standalone thing)

> require('vim.lsp.log').set_format_func(vim.inspect) helps quite a bit, I've considered making it the default

Hmm, this doesn't seem to make much/any difference for me.
The issue here is that the stderr is "just a string" from lua's perspective, but it contains lots of formatting/line breaks which humans need but vim.inspect etc escapes.

To illustrate here's a gist with the raw stderr contents (as collected by coc) vs the logs produced by nvim: https://gist.github.com/sam-mccall/849e975aad3ee4015bbc2dd51aac68a4

> It's pretty trivial to write a short lua script that will pull all [error] prefixed errors out of lsp.log, and you can ask your users to clear lsp.log before recording the issue.

We'd like users to be able to read the logs themselves, round-tripping *all* debugging through a project member who has to download the logs from the issue tracker, run them through a filtering script, and paste them back is a lot of toil/friction. Most of our project members aren't familiar with neovim or lua. Obtaining and running such a script to preprocess logs is beyond what I'd expect our users to do. (We ask them to pass `'-log=verbose`' to clangd, and only ~half do so).

**Comment by sam-mccall** [comment]
Today I got a message on startup:

`LSP client log is large (1121 MB): /usr/local/google/home/sammccall/.cache/nvim/lsp.log`

The alert is nice, but having this file accumulate logs from all sessions forever means running clangd with `-log=verbose` isn't a good idea. (In other editors this is the standard setting for clangd devs & power-users that report bugs)

**Comment by 9999years** [comment]
I would also like it to be easier to view LSP logs. I work with the haskell-language-server, which uses quite a bit of memory and takes a few minutes to start up. Until it starts, code actions and other features don't work, but there's nothing in the nvim user interface that tells me that the language server isn't ready or when it gets ready, I just have to trial-and-error hope it works or `tail -f ~/.local/state/nvim/lsp.log` in another window.

I'd love to have the LSP messages printed as Vim messages; is that possible?


> `require('vim.lsp.log').set_format_func(vim.inspect)` helps quite a bit

I've added this to my `init.lua` but it doesn't seem to change anything; what exactly is it supposed to do? I'm still seeing `\n` and similar in log messages.

**Comment by 9999years** [comment]
Also, it looks like the value sent to the function set with `vim.lsp.log.set_format_func` is a pre-formatted string that already has the timestamp and log level interpolated. Both of those are superfluous when printing the message out; the user knows the current time already. We should pass some richer object into that function instead, perhaps.

**Comment by 9999years** [comment]
I was having something of an X/Y problem here: what I really wanted was to show language server progress information. I've [achieved this](https://github.com/9999years/dotfiles/commit/ae02a8c6d9df58098d543e3f3dd42a3595821bb5) with nvim-lualine/lualine.nvim and nvim-lua/lsp-status.nvim. I still think it would be good to have better LSP logging in nvim, though.

**Comment by jackielii** [comment]
I struggled with this as well after migrating from coc.nvim to built-in lsp. I think the bottom line is there should be a "named" output/logging.

For example, we can introduce a new named logger apis:

```lua
local log = vim.lsp.log.namedLogger(lsp-client-id)
log.{debug,info,...}(...)
```

Current log interface can log to a `root` named logger.

By default, the loggers can log to a temporary buffer, which seems to be what coc.nvim does: https://github.com/neoclide/coc.nvim/blob/c5ffbb3c6b748d3359d99c4452ed74a80d293194/src/core/channels.ts#L35

User can configure the persistent log file directory. If such option exists, the buffers can persist to such directory with `{lsp-client-id}.log`, and maybe with `$NVIM_APPNAME` prefix.

`nvim-lspconfig` can use the new named logger if it's available.

What do you think? I can work on this if you agree.

**Comment by tristan957** [comment]
One use case that I would like to keep open for consideration. Today, I edited my config to spawn language servers with `systemd-run(1)`. With that and logging to stderr, I think you could use `journalctl` to view the language server logs.

**Comment by HighCommander4** [comment]
Since this issue remains unfixed and a thorn in my side as clangd maintainer, I wrote a script to undo the mess that neovim makes of language server logs:

https://github.com/HighCommander4/neovim-lsp-log-cleaner

**Comment by justinmk** [comment]
@HighCommander4 would this be resolved if we provided a filter predicate for the LSP logger? I guess "routing" also is needed. On master we have made steps towards that, but `vim.log` still needs some work.

**Comment by HighCommander4** [comment]
> [@HighCommander4](https://github.com/HighCommander4) would this be resolved if we provided a filter predicate for the LSP logger?

If I'm understanding correctly what that would mean (log lines could be filtered to just those coming from the clangd server), then no, because the cleanup script also does two other things: (1) removes prefixes added to log lines by neovim (`[ERROR][timestamp] etc.`), and (2) unescapes the contents of the lines. (2) probably makes the most difference for readability, though (1) is important too.

**Comment by justinmk** [comment]
When you say "this issue remains unfixed" I assumed you were referring to "this issue". If this is also about improving the log format, that of course is quite feasible. PR welcome...

**Comment by HighCommander4** [comment]
I understood this issue as being about making the contents of the language server's stderr available in a raw form (including e.g. no escaping). That would make the script unnecessary, since the script just tries to recover these raw contents.

**Comment by ofseed** [comment]
I may have mentioned this somewhere before: I plan to put logs from different language servers into separate files.

However, this requires a refactor of the JSON-RPC layer (so that we can distinguish different levels of logging), as well as some improvements to `vim.log` (to avoid keeping multiple files open and performing frequent writes to them at the same time).
